lib/spack/spack/package.py

```diff
@@ -1506,6 +1684,129 @@ def rpath_args(self):
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
 
+class Package(PackageBase):
+    phases = ['install']
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class EditableMakefile(PackageBase):
+    phases = ['edit', 'build', 'install']
+
+    def wdir(self):
+        return self.stage.source_path
+
+    def build_args(self):
+        return list()
+
+    def install_args(self):
+        return list()
+
+    def edit(self, spec, prefix):
+        raise NotImplementedError('\'edit\' function not implemented')
+
+    def build(self, spec, prefix):
+        args = self.build_args()
+        with working_dir(self.wdir()):
+            inspect.getmodule(self).make(*args)
+
+    def install(self, spec, prefix):
+        args = self.install_args() + ['install']
+        with working_dir(self.wdir()):
+            inspect.getmodule(self).make(*args)
+
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class AutotoolsPackage(PackageBase):
+    phases = ['autoreconf', 'configure', 'build', 'install']
+
+    def autoreconf(self, spec, prefix):
+        """Not needed usually, configure should be already there"""
+        pass
+
+    @PackageBase.sanity_check('autoreconf')
+    def is_configure_or_die(self):
+        if not os.path.exists('configure'):
+            raise RuntimeError(
+                'configure script not found in {0}'.format(os.getcwd()))
+
+    def configure_args(self):
```

Comment:
Actually... I think it shouldn't be.  It will be more of a pain for package authors to write `@property` all over the place; and they won't understand why.